quickwit/quickwit-serve/src/rest.rs

```diff
@@ -107,14 +109,24 @@ pub(crate) async fn start_rest_server(
         .and(warp::get())
         .map(|| redirect(http::Uri::from_static("/ui/search")));
 
-    // Some users run dashboards, etc... On different ports
-    // to the one Quickwit uses, this causes an error with the browsers
-    // For now we can allow all origins since Quickwit needs to be
-    // behind a reverse proxy to be publicly exposed currently
-    // but in future we may want to change this.
-    let cors = warp::cors()
-        .allow_any_origin()
-        .allow_methods(["GET", "POST", "PUT", "DELETE"]);
+    let mut cors =
+        CorsLayer::new().allow_methods([Method::GET, Method::POST, Method::PUT, Method::DELETE]);
+    let cors_origins = &quickwit_services.config.rest_config.cors_allow_origins;
+    if !cors_origins.is_empty() {
+        let allow_any = cors_origins.iter().any(|origin| origin.as_str() == "*");
+
+        if allow_any {
+            info!("CORS is enabled, all origins will be allowed");
+            cors = cors.allow_origin(tower_http::cors::Any);
+        } else {
+            info!(origins = ?cors_origins, "CORS is enabled, the following origins will be allowed");
+            let origins = cors_origins
+                .iter()
+                .map(|origin| origin.parse::<HeaderValue>().unwrap())
+                .collect::<Vec<_>>();
+            cors = cors.allow_origin(origins);
+        };
+    }
```

Comment:
Warp's CORS middleware did not seem to allow you to specify origins *and not* reject any request with an invalid origin. This is why we went with Tower so that we do enough to satisfy the browsers without interfering with the CLI or other applications.